src/frontend/src/optimizer/plan_node/logical_over_agg.rs

```diff
@@ -153,75 +162,115 @@ impl LogicalOverAgg {
+        // in `LogicalAgg`
+        let order_by: Vec<_> = window_function
+            .order_by
             .sort_exprs
             .into_iter()
             .map(|e| match e.expr.as_input_ref() {
                 Some(i) => Ok(ColumnOrder::new(i.index(), e.order_type)),
                 None => Err(ErrorCode::NotImplemented(
                     "ORDER BY expression in window function".to_string(),
                     None.into(),
-                )
-                .into()),
+                )),
             })
-            .collect::<Result<Vec<_>>>()?;
-        let partition_by = partition_by
+            .try_collect()?;
+        let partition_by: Vec<_> = window_function
+            .partition_by
             .into_iter()
             .map(|e| match e.as_input_ref() {
                 Some(i) => Ok(*i.clone()),
                 None => Err(ErrorCode::NotImplemented(
                     "PARTITION BY expression in window function".to_string(),
                     None.into(),
-                )
-                .into()),
+                )),
             })
-            .collect::<Result<Vec<_>>>()?;
-
-        let over_agg = Self::new(
-            PlanWindowFunction {
-                function_type,
-                return_type,
-                partition_by,
-                order_by,
-            },
-            input,
-        );
-        Ok((over_agg.into(), select_exprs))
+            .try_collect()?;
+
+        let mut args = window_function.args;
+        let frame = match window_function.kind {
+            WindowFuncKind::RowNumber | WindowFuncKind::Rank | WindowFuncKind::DenseRank => {
+                // ignore frame for rank functions
+                None
+            }
+            WindowFuncKind::Lag | WindowFuncKind::Lead => {
+                let offset = if args.len() > 1 {
+                    let offset_expr = args.remove(1);
+                    if !offset_expr.return_type().is_int() {
+                        return Err(ErrorCode::InvalidInputSyntax(format!(
+                            "the `offset` of `{}` function should be integer",
+                            window_function.kind
+                        ))
+                        .into());
+                    }
+                    offset_expr
+                        .cast_implicit(DataType::Int64)?
+                        .eval_row_const()?
+                        .map(|v| *v.as_int64() as usize)
+                        .unwrap_or(1usize)
+                } else {
+                    1usize
+                };
+
+                // override the frame
+                // TODO(rc): We can only do the optimization for constant offset.
+                Some(if window_function.kind == WindowFuncKind::Lag {
+                    Frame::Rows(FrameBound::Preceding(offset), FrameBound::CurrentRow)
+                } else {
+                    Frame::Rows(FrameBound::CurrentRow, FrameBound::Following(offset))
+                })
```

Comment:
LogicalWindow shouldn't override the original SQL. This limitation is all about `StreamWindow`, so we can detect it in the `StreamWindow` and then throw an error.

Reply:
Agree, will fix this in next PR.

Reply:
Actually the frame of `lag` and `lead` is also ignore in PG, so if we move the optimization for constant offset where we generate `StreamOverWindow`, we still have to decide whether to override `window_function.frame` here or later.

Reply:
Seems `Rank`, `Lead`, and `Lag` wouldn't be affected by the frame. 🥵